src/CircuitZKit.ts

```diff
@@ -0,0 +1,254 @@
+  }
+
+  public async generateCalldata(proof: ProofStruct): Promise<Calldata> {
+    const calldata = await snarkjs.groth16.exportSolidityCallData(proof.proof, proof.publicSignals);
+
+    return JSON.parse(`[${calldata}]`) as Calldata;
+  }
+
+  public getCircuitId(): string {
+    return path.parse(this._circuit).name;
+  }
+
+  public getVerifierId(): string {
+    return `${path.parse(this._circuit).name}Verifier`;
+  }
+
+  private async _generateZKey(outDir: string): Promise<void> {
+    const r1csFile = this._getFile("r1cs", outDir);
+    const zKeyFile = this._getFile("zkey", outDir);
+
+    const constraints = await this._getConstraints(outDir);
+    const ptauInfo = await this._manager.fetchPtauInfo(constraints);
+
+    await snarkjs.zKey.newZKey(r1csFile, ptauInfo.file, zKeyFile);
+  }
+
+  private async _generateVKey(outDir: string): Promise<void> {
+    const zKeyFile = this._getFile("zkey", outDir);
+    const vKeyFile = this._getFile("vkey", outDir);
+
+    const vKeyData = await snarkjs.zKey.exportVerificationKey(zKeyFile);
+
+    fs.writeFileSync(vKeyFile, JSON.stringify(vKeyData));
+  }
+
+  private _getCompileArgs(options: CompileOptions, outDir: string): string[] {
+    let args = [this._circuit, "--r1cs", "--wasm"];
+
+    options.sym && args.push("--sym");
+    options.json && args.push("--json");
+    options.c && args.push("--c");
+    !options.quiet && args.push("--verbose");
+
+    args.push("-o", outDir);
+
+    return args;
+  }
+
+  private async _compile(options: CompileOptions, outDir: string): Promise<void> {
+    const args = this._getCompileArgs(options, outDir);
+
+    try {
+      await this._getCircomRunner(args, options.quiet).execute(this._manager.getCompiler());
+    } catch (err) {
+      if (options.quiet) {
+        throw new Error(
+          'Compilation failed with an unknown error. Consider passing "quiet=false" flag to see the compilation error.',
+          { cause: err },
+        );
+      }
+
+      throw new Error("Compilation failed.", { cause: err });
+    }
+  }
+
+  async _getConstraints(outDir: string): Promise<number> {
+    const r1csFile = this._getFile("r1cs", outDir);
+
+    const r1csDescriptor = fs.openSync(r1csFile, "r");
+
+    const readBytes = (position: number, length: number): bigint => {
+      const buffer = Buffer.alloc(length);
+
+      fs.readSync(r1csDescriptor, buffer, { length, position });
+
+      return BigInt(`0x${buffer.reverse().toString("hex")}`);
+    };
+
+    /// @dev https://github.com/iden3/r1csfile/blob/master/doc/r1cs_bin_format.md#format-of-the-file
+    const numberOfSections = readBytes(8, 4);
+    let sectionStart = 12;
+
+    for (let i = 0; i < numberOfSections; ++i) {
+      const sectionType = Number(readBytes(sectionStart, 4));
+      const sectionSize = Number(readBytes(sectionStart + 4, 8));
+
+      /// @dev Reading header section
+      if (sectionType == 1) {
+        const totalConstraintsOffset = 4 + 8 + 4 + 32 + 4 + 4 + 4 + 4 + 8;
```

Comment:
Maybe a separate file called constants.ts should be created to store these value

Reply:
I think there should not be a global (changeable) constant for this thing